# Task
Phân tích tệp `financial_fraud_detection_dataset.csv` để phát hiện gian lận bằng cách tải dữ liệu vào DataFrame của pandas, khám phá cấu trúc, tiền xử lý (bao gồm xử lý đặc trưng phân loại và giá trị thiếu), chia tách dữ liệu thành tập huấn luyện và kiểm tra, chuẩn hóa đặc trưng số, huấn luyện mô hình mạng nơ-ron (ví dụ: `MLPClassifier` từ scikit-learn), và đánh giá hiệu suất mô hình bằng các chỉ số như độ chính xác, độ nhạy, độ chính xác và điểm F1. Cuối cùng, tóm tắt và giải thích chi tiết tất cả các bước đã thực hiện và kết quả đánh giá mô hình bằng tiếng Việt.

## Tải và Khám phá dữ liệu

### Subtask:
Tải tệp `financial_fraud_detection_dataset.csv` vào một DataFrame của pandas và khám phá cấu trúc dữ liệu, các cột, loại dữ liệu, và kiểm tra giá trị thiếu. Bước này giúp hiểu rõ hơn về dữ liệu đầu vào.


**Reasoning**:
To begin exploring the dataset, I need to import the pandas library, load the CSV file into a DataFrame, display its head, print its information summary, and check for missing values as per the instructions.



In [ ]:
import pandas as pd

# Load the CSV file into a DataFrame
df = pd.read_csv('/content/financial_fraud_detection_dataset.csv')

print("First 5 rows of the DataFrame:")
print(df.head())

print("\nDataFrame Information:")
df.info()

print("\nNumber of missing values per column:")
print(df.isnull().sum())

First 5 rows of the DataFrame:
  transaction_id                   timestamp sender_account receiver_account  \
0        T100000  2023-08-22T09:22:43.516168      ACC877572        ACC388389   
1        T100001  2023-08-04T01:58:02.606711      ACC895667        ACC944962   
2        T100002  2023-05-12T11:39:33.742963      ACC733052        ACC377370   
3        T100003  2023-10-10T06:04:43.195112      ACC996865        ACC344098   
4        T100004  2023-09-24T08:09:02.700162      ACC584714        ACC497887   

    amount transaction_type merchant_category location device_used  is_fraud  \
0   343.78       withdrawal         utilities    Tokyo      mobile     False   
1   419.65       withdrawal            online  Toronto         atm     False   
2  2773.86          deposit             other   London         pos     False   
3  1666.22          deposit            online   Sydney         pos     False   
4    24.43         transfer         utilities  Toronto      mobile     False   

  fraud

## Tiền xử lý dữ liệu

### Subtask:
Chuẩn bị dữ liệu cho việc huấn luyện mô hình. Điều này bao gồm việc xử lý các đặc trưng phân loại (ví dụ: mã hóa one-hot), xử lý các giá trị thiếu nếu có, và chia tách dữ liệu thành tập đặc trưng (X) và biến mục tiêu (y).


**Reasoning**:
The first step in data preprocessing is to remove columns that are not necessary for modeling or have too many unique values that could lead to high dimensionality or are directly related to the target variable, as per the instructions. This includes 'transaction_id', 'sender_account', 'receiver_account', 'ip_address', 'device_hash', and 'fraud_type'.



In [ ]:
columns_to_drop = ['transaction_id', 'sender_account', 'receiver_account', 'ip_address', 'device_hash', 'fraud_type']
df_processed = df.drop(columns=columns_to_drop)

print(f"Columns dropped: {columns_to_drop}")
print("Shape of DataFrame after dropping columns:", df_processed.shape)

Columns dropped: ['transaction_id', 'sender_account', 'receiver_account', 'ip_address', 'device_hash', 'fraud_type']
Shape of DataFrame after dropping columns: (251898, 12)


**Reasoning**:
The next step is to convert the 'timestamp' column to datetime objects and then extract new features such as 'hour', 'day_of_week', and 'month' from it, as specified in the instructions. Finally, the original 'timestamp' column will be dropped.



In [ ]:
df_processed['timestamp'] = pd.to_datetime(df_processed['timestamp'])
df_processed['hour'] = df_processed['timestamp'].dt.hour
df_processed['day_of_week'] = df_processed['timestamp'].dt.dayofweek
df_processed['month'] = df_processed['timestamp'].dt.month
df_processed = df_processed.drop(columns=['timestamp'])

print("DataFrame after extracting time features and dropping 'timestamp' column:")
print(df_processed.head())
print("Shape of DataFrame after creating new time features:", df_processed.shape)

DataFrame after extracting time features and dropping 'timestamp' column:
    amount transaction_type merchant_category location device_used  is_fraud  \
0   343.78       withdrawal         utilities    Tokyo      mobile     False   
1   419.65       withdrawal            online  Toronto         atm     False   
2  2773.86          deposit             other   London         pos     False   
3  1666.22          deposit            online   Sydney         pos     False   
4    24.43         transfer         utilities  Toronto      mobile     False   

   time_since_last_transaction  spending_deviation_score  velocity_score  \
0                          NaN                     -0.21               3   
1                          NaN                     -0.14               7   
2                          NaN                     -1.78              20   
3                          NaN                     -0.60               6   
4                          NaN                      0.79         

**Reasoning**:
Following the instructions, I need to handle missing values in the 'time_since_last_transaction' column by filling them with 0. This is based on the assumption that a missing value in this column indicates either the first transaction or a long period since the last one.



In [ ]:
df_processed['time_since_last_transaction'].fillna(0, inplace=True)

print("Number of missing values in 'time_since_last_transaction' after imputation:")
print(df_processed['time_since_last_transaction'].isnull().sum())
print("First 5 rows of DataFrame after handling missing values:")
print(df_processed.head())

Number of missing values in 'time_since_last_transaction' after imputation:
0
First 5 rows of DataFrame after handling missing values:
    amount transaction_type merchant_category location device_used  is_fraud  \
0   343.78       withdrawal         utilities    Tokyo      mobile     False   
1   419.65       withdrawal            online  Toronto         atm     False   
2  2773.86          deposit             other   London         pos     False   
3  1666.22          deposit            online   Sydney         pos     False   
4    24.43         transfer         utilities  Toronto      mobile     False   

   time_since_last_transaction  spending_deviation_score  velocity_score  \
0                          0.0                     -0.21               3   
1                          0.0                     -0.14               7   
2                          0.0                     -1.78              20   
3                          0.0                     -0.60               6   
4   

/tmp/ipython-input-3920246974.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_processed['time_since_last_transaction'].fillna(0, inplace=True)


**Reasoning**:
The previous step produced a `FutureWarning` due to the use of `inplace=True` with chained assignment. To fix this, I will modify the line to directly assign the result of `fillna(0)` back to the column, which is the recommended practice for avoiding such warnings and ensuring the operation is performed on the DataFrame.



In [ ]:
df_processed['time_since_last_transaction'] = df_processed['time_since_last_transaction'].fillna(0)

print("Number of missing values in 'time_since_last_transaction' after imputation:")
print(df_processed['time_since_last_transaction'].isnull().sum())
print("First 5 rows of DataFrame after handling missing values:")
print(df_processed.head())

Number of missing values in 'time_since_last_transaction' after imputation:
0
First 5 rows of DataFrame after handling missing values:
    amount transaction_type merchant_category location device_used  is_fraud  \
0   343.78       withdrawal         utilities    Tokyo      mobile     False   
1   419.65       withdrawal            online  Toronto         atm     False   
2  2773.86          deposit             other   London         pos     False   
3  1666.22          deposit            online   Sydney         pos     False   
4    24.43         transfer         utilities  Toronto      mobile     False   

   time_since_last_transaction  spending_deviation_score  velocity_score  \
0                          0.0                     -0.21               3   
1                          0.0                     -0.14               7   
2                          0.0                     -1.78              20   
3                          0.0                     -0.60               6   
4   

**Reasoning**:
Now that missing values are handled, I will define the target variable `y` as 'is_fraud' and the feature set `X` as all other columns from `df_processed`. This step is crucial for preparing the data for model training.



In [ ]:
y = df_processed['is_fraud']
X = df_processed.drop(columns=['is_fraud'])

print("Shape of X (features):", X.shape)
print("Shape of y (target):", y.shape)
print("First 5 rows of y:")
print(y.head())

Shape of X (features): (251898, 13)
Shape of y (target): (251898,)
First 5 rows of y:
0    False
1    False
2    False
3    False
4    False
Name: is_fraud, dtype: bool


**Reasoning**:
The next step in data preprocessing is to apply one-hot encoding to all categorical columns (columns with 'object' dtype) in the feature set `X` using `pd.get_dummies()`, as specified in the instructions. This will convert the categorical features into a numerical format suitable for machine learning models.



In [ ]:
X_processed = pd.get_dummies(X, drop_first=True)

print("Shape of X after one-hot encoding:", X_processed.shape)
print("First 5 rows of X after one-hot encoding:")
print(X_processed.head())

Shape of X after one-hot encoding: (251898, 31)
First 5 rows of X after one-hot encoding:
    amount  time_since_last_transaction  spending_deviation_score  \
0   343.78                          0.0                     -0.21   
1   419.65                          0.0                     -0.14   
2  2773.86                          0.0                     -1.78   
3  1666.22                          0.0                     -0.60   
4    24.43                          0.0                      0.79   

   velocity_score  geo_anomaly_score  hour  day_of_week  month  \
0               3               0.22     9            1      8   
1               7               0.96     1            4      8   
2              20               0.89    11            4      5   
3               6               0.37     6            1     10   
4              13               0.27     8            6      9   

   transaction_type_payment  transaction_type_transfer  ...  \
0                     False        

## Chia tách và Chuẩn hóa dữ liệu

### Subtask:
Chia dữ liệu đã tiền xử lý thành các tập huấn luyện và kiểm tra. Chuẩn hóa (scaling) các đặc trưng số để tối ưu hóa hiệu suất của mô hình và đảm bảo tất cả các đặc trưng đóng góp tương đương vào quá trình huấn luyện.


**Reasoning**:
The first step is to split the preprocessed data into training and testing sets using `train_test_split` as instructed, which is crucial for evaluating the model's performance on unseen data.



In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_processed, y, test_size=0.2, random_state=42)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (201518, 31)
Shape of X_test: (50380, 31)
Shape of y_train: (201518,)
Shape of y_test: (50380,)


**Reasoning**:
Now that the data is split into training and testing sets, I need to standardize the numerical features to ensure all features contribute equally to the model training, as per the instructions. This involves identifying numerical columns, initializing a `StandardScaler`, fitting it on the numerical columns of the training set, and then transforming both the training and test sets.



In [ ]:
from sklearn.preprocessing import StandardScaler

# Identify numerical columns (those not created by one-hot encoding)
# By checking the dtypes, original numerical columns are float64 or int64 before one-hot encoding.
# After one-hot encoding, they retain their types, while new columns are boolean.
# We can assume numerical features are those that are not boolean.
numerical_cols = X_train.select_dtypes(include=['int64', 'float64']).columns

# Initialize StandardScaler
scaler = StandardScaler()

# Fit the scaler on the numerical columns of the training data
X_train[numerical_cols] = scaler.fit_transform(X_train[numerical_cols])

# Transform the numerical columns of both training and test data
X_test[numerical_cols] = scaler.transform(X_test[numerical_cols])

print("First 5 rows of X_train after scaling numerical features:")
print(X_train.head())
print("\nFirst 5 rows of X_test after scaling numerical features:")
print(X_test.head())

First 5 rows of X_train after scaling numerical features:
          amount  time_since_last_transaction  spending_deviation_score  \
27542  -0.372807                     0.000166                 -1.694139   
171173  0.242596                     0.000166                 -0.102541   
66303   0.163053                     0.000166                  0.047609   
10474  -0.747608                     0.000166                 -0.983426   
31862  -0.329426                     0.000166                 -1.794240   

        velocity_score  geo_anomaly_score  hour  day_of_week  month  \
27542        -0.780948           1.006859    11            0      1   
171173        0.779988           1.180182     5            0      5   
66303         0.953425          -1.419673    18            0      5   
10474         0.433113          -0.102413    13            5     11   
31862        -0.607511           0.278899     7            5      6   

        transaction_type_payment  transaction_type_transfer  ...

## Huấn luyện Mô hình Mạng Nơ-ron

### Subtask:
Xây dựng và huấn luyện một mô hình mạng nơ-ron đơn giản (ví dụ: `MLPClassifier` từ thư viện scikit-learn, đại diện cho một mô hình mạng nơ-ron cho dữ liệu bảng) trên tập dữ liệu huấn luyện. Giải thích rằng mô hình này được chọn vì "HTGNN" không phải là một mô hình tiêu chuẩn.


**Reasoning**:
I need to import `MLPClassifier`, initialize it with specified parameters, and then fit it on the `X_train` and `y_train` data as per the instructions. Finally, I should print a confirmation message.



In [ ]:
from sklearn.neural_network import MLPClassifier

# Initialize the MLPClassifier model
# Using hidden_layer_sizes=(100,), max_iter=100, and random_state=42 as initial parameters
mlp = MLPClassifier(hidden_layer_sizes=(100,), max_iter=100, random_state=42, verbose=True)

# Train the model
print("Training MLPClassifier...")
mlp.fit(X_train, y_train)

print("MLPClassifier training complete!")

Training MLPClassifier...
Iteration 1, loss = 0.03537676
Iteration 2, loss = 0.03101891
Iteration 3, loss = 0.03009368
Iteration 4, loss = 0.02959512
Iteration 5, loss = 0.02919465
Iteration 6, loss = 0.02863791
Iteration 7, loss = 0.02818207
Iteration 8, loss = 0.02789934
Iteration 9, loss = 0.02746690
Iteration 10, loss = 0.02708533
Iteration 11, loss = 0.02692524
Iteration 12, loss = 0.02659947
Iteration 13, loss = 0.02646306
Iteration 14, loss = 0.02616153
Iteration 15, loss = 0.02599342
Iteration 16, loss = 0.02587097
Iteration 17, loss = 0.02567453
Iteration 18, loss = 0.02556704
Iteration 19, loss = 0.02541843
Iteration 20, loss = 0.02517416
Iteration 21, loss = 0.02503104
Iteration 22, loss = 0.02491958
Iteration 23, loss = 0.02497839
Iteration 24, loss = 0.02471233
Iteration 25, loss = 0.02461736
Iteration 26, loss = 0.02435466
Iteration 27, loss = 0.02441557
Iteration 28, loss = 0.02423349
Iteration 29, loss = 0.02416855
Iteration 30, loss = 0.02400673
Iteration 31, loss = 0.

## Đánh giá hiệu suất mô hình

### Subtask:
Đánh giá hiệu suất của mô hình mạng nơ-ron đã huấn luyện trên tập kiểm tra (`X_test`, `y_test`) bằng cách tính toán các chỉ số như độ chính xác (accuracy), độ nhạy (recall), độ chính xác (precision) và điểm F1. Các chỉ số này sẽ giúp đánh giá mức độ hiệu quả của mô hình trong việc phát hiện gian lận.


**Reasoning**:
Now that the model has been trained, I need to predict on the test set and calculate evaluation metrics such as accuracy, precision, recall, and F1-score to assess its performance as specified in the subtask.



In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Make predictions on the test set
y_pred = mlp.predict(X_test)

# Calculate evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print(f"Model Evaluation Metrics:\n")
print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-Score: {f1:.4f}")

Model Evaluation Metrics:

Accuracy: 0.9933
Precision: 0.0444
Recall: 0.0068
F1-Score: 0.0117


## Hoàn thành tác vụ

### Subtask:
Tóm tắt kết quả đánh giá mô hình và cung cấp giải thích chi tiết về các bước đã thực hiện, tất cả bằng tiếng Việt. Giải thích rằng một mạng nơ-ron tổng quát đã được sử dụng và nếu "HTGNN" có ý nghĩa cụ thể liên quan đến đồ thị hoặc kiến trúc đặc biệt, cần cung cấp thêm thông tin để tùy chỉnh mô hình.


## Summary:

Dưới đây là tóm tắt quá trình giải quyết nhiệm vụ:

### Q&A
Nhiệm vụ yêu cầu tóm tắt kết quả đánh giá mô hình và cung cấp giải thích chi tiết về các bước đã thực hiện. Một mạng nơ-ron tổng quát đã được sử dụng (MLPClassifier). Nếu "HTGNN" có ý nghĩa cụ thể liên quan đến đồ thị hoặc kiến trúc đặc biệt, cần cung cấp thêm thông tin để tùy chỉnh mô hình. Trong trường hợp này, "HTGNN" không được định nghĩa rõ ràng, nên mô hình MLPClassifier đã được chọn làm đại diện cho mạng nơ-ron tổng quát.

### Data Analysis Key Findings
*   **Dữ liệu ban đầu**: Tập dữ liệu `financial_fraud_detection_dataset.csv` chứa 251.898 giao dịch với 18 cột.
    *   Các cột `fraud_type` và `time_since_last_transaction` có số lượng giá trị thiếu đáng kể (250.501 và 219.838 tương ứng).
*   **Tiền xử lý dữ liệu**:
    *   Các cột không cần thiết (`transaction_id`, `sender_account`, `receiver_account`, `ip_address`, `device_hash`, `fraud_type`) đã được loại bỏ.
    *   Đã trích xuất các đặc trưng thời gian (`hour`, `day_of_week`, `month`) từ cột `timestamp`.
    *   Các giá trị thiếu trong `time_since_last_transaction` đã được điền bằng 0.
    *   Dữ liệu được chia thành tập đặc trưng (X) và biến mục tiêu (y - `is_fraud`).
    *   Các đặc trưng phân loại đã được mã hóa one-hot, dẫn đến tập đặc trưng `X_processed` có kích thước (251898, 31).
*   **Chia tách và Chuẩn hóa dữ liệu**:
    *   Dữ liệu được chia thành tập huấn luyện (80%, 201.518 mẫu) và tập kiểm tra (20%, 50.380 mẫu) với `random_state=42`.
    *   Các đặc trưng số đã được chuẩn hóa bằng `StandardScaler`.
*   **Huấn luyện mô hình**: Một mô hình `MLPClassifier` với `hidden_layer_sizes=(100,)` và `max_iter=100` đã được huấn luyện thành công trên tập dữ liệu huấn luyện.
*   **Đánh giá hiệu suất mô hình**: Trên tập kiểm tra, mô hình đạt được các chỉ số sau:
    *   Độ chính xác (Accuracy): 0.9933
    *   Độ chính xác (Precision) cho lớp gian lận: 0.0444
    *   Độ nhạy (Recall) cho lớp gian lận: 0.0068
    *   Điểm F1 (F1-Score) cho lớp gian lận: 0.0117

### Insights or Next Steps
*   Mặc dù mô hình đạt độ chính xác tổng thể cao (99.33%), các chỉ số `Precision`, `Recall` và `F1-Score` cực kỳ thấp cho việc phát hiện giao dịch gian lận. Điều này cho thấy mô hình không hiệu quả trong việc xác định các trường hợp gian lận thực tế, có thể là do sự mất cân bằng dữ liệu nghiêm trọng (số lượng giao dịch gian lận rất ít).
*   Để cải thiện hiệu suất phát hiện gian lận, cần áp dụng các kỹ thuật xử lý dữ liệu mất cân bằng (ví dụ: Oversampling, Undersampling, SMOTE) hoặc sử dụng các thuật toán nhận biết mất cân bằng dữ liệu, và tinh chỉnh các tham số của `MLPClassifier` hoặc thử các mô hình khác phù hợp hơn với dữ liệu mất cân bằng.
